### ISSUE TITLE
For classification operators, add two synthetic features: class predictions and class probabilities

### ISSUE THREAD

**Issue by rhiever** [rhiever/tpot#214]
Whenever TPOT runs through a classification operator, it adds the class labels predicted by the classifier as a synthetic feature. Let's also add _N_ additional synthetic features, which are the class probabilities predicted by `predict_proba`, if that is available for the given classifier. It's possible that such synthetic features will be useful for other classifiers down the pipeline.

**Comment by teaearlgraycold** [comment]
Not sure if this is possible when using sklearn pipelines.

**Comment by rhiever** [comment]
What if we wrapped the probabilistic predictor with a FunctionTransformer?

**Comment by teaearlgraycold** [comment]
Do you have an example of that in a pipeline? I'm not sure what that would look like.

**Comment by rhiever** [comment]
I tried to demo this but couldn't figure out how to get a regular pipeline (that's using `predict`) to call `predict_prob` on an internal pipeline. Ideas @amueller?

**Comment by amueller** [comment]
I'm not sure what you mean.

**Comment by teaearlgraycold** [comment]
As I understand it, we need something in a pipeline that will call fit and subsequently predict_proba on an estimator. But pipelines are designed to call fit and then predict.

**Comment by rhiever** [comment]
We're basically looking for a way to do, e.g.,

(hacky code incoming)

```
Pipeline(FeatureUnion(RandomForestClassifier().predict(), RandomForestClassifier().predict_proba(), FunctionTransformer(lambda x: return x)), GaussianNB())
```

So every time there's a nested classifier within the pipeline (e.g. Data --> RF --> GaussianNB, as above), we pass to the next operator a copy of the current features with new features appended: the predicted classes from that nested classifier, and the predicted class probabilities from that nested classifier.

**Comment by amueller** [comment]
VotingClassifier?

**Comment by rhiever** [comment]
I've almost got it working, except that `VotingClassifier.transform` returns a matrix of shape `(1, # samples, # features)`. What's up with that? Any workaround?

```python
from sklearn.ensemble import RandomForestClassifier, VotingClassifier
from sklearn.pipeline import make_pipeline, make_union
from sklearn.preprocessing import FunctionTransformer
from sklearn.datasets import make_classification

features, targets = make_classification(n_classes=10, n_informative=20, n_features=100, n_samples=1000)

def identity(x):
    return x

clf = make_pipeline(make_union(VotingClassifier([('rf1', RandomForestClassifier())], voting='soft'),
                               FunctionTransformer(identity)),
                    RandomForestClassifier())
```

**Comment by amueller** [comment]
If `voting=’soft’`: :

    array-like = [n_classifiers, n_samples, n_classes]

        Class probabilities calculated by each classifier.

Maybe add a "flatten" parameter to VotingClassifier?

Obvious workaround for now is FunctionTransformer :-/

**Comment by amueller** [comment]
https://github.com/scikit-learn/scikit-learn/issues/7230

**Comment by rhiever** [comment]
>Obvious workaround for now is FunctionTransformer :-/

That was my thought too. Though I think I'm too tired right now to figure out how to apply the FunctionTransformer right after the soft VotingClassifier.

**Comment by amueller** [comment]
pipeline?

**Comment by rhiever** [comment]
FunctionTransformers only have transform. Pipelines require a predict at the end. Need some sort of "TransformerPipeline".

**Comment by amueller** [comment]
no, pipelines don't require a predict at the end.

**Comment by rhiever** [comment]
Oh duh, I knew that. Was clearly over tired yesterday.

**Comment by rhiever** [comment]
Hmmm, no dice. The pipeline complains about the 3-dimensional output before even passing it to the FunctionTransformer:

```python
from sklearn.ensemble import RandomForestClassifier, VotingClassifier
from sklearn.pipeline import make_pipeline, make_union
from sklearn.preprocessing import FunctionTransformer
from sklearn.datasets import make_classification

features, targets = make_classification(n_classes=10, n_informative=20, n_features=100, n_samples=1000)

def identity(x):
    return x

def subset(x):
    print(x)
    return x[0]

clf = make_pipeline(make_union(make_pipeline(
            VotingClassifier([('rf1', RandomForestClassifier())], voting='soft'),
            FunctionTransformer(subset)),
                               VotingClassifier([('rf1', RandomForestClassifier())], voting='hard'),
                               FunctionTransformer(identity)),
                    RandomForestClassifier())
```

**Comment by amueller** [comment]
http://sscce.org/

**Comment by absolutelyNoWarranty** [comment]
@rhiever 
Pass `validate=False` to FunctionTransformer

**Comment by rhiever** [comment]
Awesome, you're right @absolutelyNoWarranty! The code below works:

```python
from sklearn.ensemble import RandomForestClassifier, VotingClassifier
from sklearn.pipeline import make_pipeline, make_union
from sklearn.preprocessing import FunctionTransformer
from sklearn.datasets import make_classification

features, targets = make_classification(n_classes=10, n_informative=20, n_features=100, n_samples=1000)

def identity(x):
    return x

def subset(x):
    return x[0]

clf = make_pipeline(make_union(make_pipeline(
            VotingClassifier([('rf1', RandomForestClassifier())], voting='soft'),
            FunctionTransformer(subset, validate=False)),
                               VotingClassifier([('rf1', RandomForestClassifier())], voting='hard'),
                               FunctionTransformer(identity)),
                    RandomForestClassifier())

clf.fit(features, targets)
print(clf.score(features, targets))
```

Extra proof:

```python
from sklearn.ensemble import RandomForestClassifier, VotingClassifier
from sklearn.pipeline import make_pipeline, make_union
from sklearn.preprocessing import FunctionTransformer
from sklearn.datasets import make_classification

features, targets = make_classification(n_classes=10, n_informative=20, n_features=100, n_samples=1000)

def subset(x):
    return x[0]

clf = make_pipeline(VotingClassifier([('rf1', RandomForestClassifier())], voting='soft'),
                    FunctionTransformer(subset, validate=False))

clf.fit(features, targets)
print(clf.transform(features).shape)
```

cc @amueller

**Comment by amueller** [comment]
well, yeah, obv. You said "Before passing it to function transformer" which seemed wrong to me, which is why I asked for an example ;)

**Comment by rhiever** [comment]
It'd be ideal if it happened before passing it to FunctionTransformer, but this way works in the meantime. :-)
